Mathlib/CategoryTheory/ConcreteCategory/Pointed.lean

```diff
@@ -0,0 +1,177 @@
+
+* `Pointed.map`, for a map of concrete categories, takes the pointed objects of `C` to
+  the pointed objects of `D`.
+
+* `Pointed.functor` gives, for a map of concrete categories, a functor between the correponding
+  categories of pointed objects.
+
+### Implementation notes
+
+This construction is equivalent to a special case of the category of elements construction and
+also to a particular comma construction. This files provides a more convenient API for directly
+working with the category of pointed objects. We prove the equivalences in
+`CategoryTheory.ConcreteCategory.Pointed.elementsEquivalence` and
+`CategoryTheory.ConcreteCategory.structuredArrowEquivalence`.
+-/
+
+namespace CategoryTheory
+
+universe w v u
+
+variable {C : Type u} [Category.{v} C] [ConcreteCategory.{w,v,u} C]
+
+namespace ConcreteCategory
+
+attribute [local instance] ConcreteCategory.hasCoeToSort
+attribute [local instance] ConcreteCategory.instFunLike
+
+variable (C) in
+
+/-- The pointed objects of a category `C`. A pointed object in `C` is an object `X : C`
+together with a distinguished element `x` of the underlying type of `X`, often called
+the basepoint. -/
+class Pointed where
+  /-- the underlying object -/
+  obj : C
+  /-- the basepoint -/
+  pt : obj
+
+namespace Pointed
+
+@[ext]
+lemma ext {A B : Pointed C} (h₁ : A.obj = B.obj)
+    (h₂ : ConcreteCategory.forget.map (eqToHom h₁) A.pt = B.pt) :
+    A = B := by
+  obtain ⟨X, x⟩ := A
+  obtain ⟨Y, y⟩ := B
+  cases h₁
+  congr 1
+  aesop
+
+/-- The type of morphisms between pointed objects. A morphism of pointed objects is a morphism
+in the original category between the objects which preserve the basepoints. -/
+structure Hom (A B : Pointed C) where
+  /-- the object morphism -/
+  obj : A.obj ⟶ B.obj
+  /-- compatibility with the basedpoint -/
+  pt : obj A.pt = B.pt := by aesop_cat
+
+/-- The category structure on the category of pointed objects. -/
+@[simps!]
+instance category : Category (Pointed C) where
+  Hom := Pointed.Hom
+  id A := ⟨𝟙 A.obj, by aesop_cat⟩
+  comp {X Y Z} f g := ⟨f.obj ≫ g.obj, by
+    simp [f.pt, g.pt]⟩
+
+/-- Constructor for morphisms of the category pointed objects. -/
+@[simps!]
+def homMk (A B : Pointed C) (f : A.obj ⟶ B.obj) (hf : f A.pt = B.pt) :
+    A ⟶ B :=
+  ⟨f, hf⟩
+
+@[ext]
+theorem Hom.ext {A B : Pointed C} (f g : A ⟶ B) (w : f.obj = g.obj) : f = g := by
+  cases f
+  cases g
+  cases w
+  rfl
+
+@[simp]
+theorem comp_obj {A B C : Pointed C} {f : A ⟶ B} {g : B ⟶ C} : (f ≫ g).obj = f.obj ≫ g.obj := rfl
+
+@[simp]
+theorem id_obj {A : Pointed C} : (𝟙 A : A ⟶ A).obj = 𝟙 A.obj := rfl
+
+@[simp]
+theorem id_pt {A B : Pointed C} (f : A ⟶ B) : f.obj A.pt = B.pt := f.pt
+
+instance concreteCategory : ConcreteCategory (Pointed C) where
+  forget :=
+    { obj := fun A ↦ A.obj
+      map := fun f ↦ f.obj }
+  forget_faithful := by
+    refine ⟨@fun X Y ↦ ?_⟩
+    dsimp
+    fapply Function.Injective.comp
+    · apply forget_faithful.map_injective
+    · apply Hom.ext
```

Comment:
I imagine it would be useful to define a "forgetful" functor `Pointed C ⥤ C` and prove that it is faithful. 
Then you can define the forgetful functor from `Pointed C` as the composition of that with the forgetful functor from `C`, and faithfulness will be inferred automatically.